cpgserver/src/main/scala/io/shiftleft/cpgservernew/route/CpgProvider.scala

```diff
@@ -0,0 +1,31 @@
+package io.shiftleft.cpgservernew.route
+
+import java.util.UUID
+
+import cats.data.OptionT
+import cats.effect.IO
+
+import io.shiftleft.codepropertygraph.Cpg
+
+trait CpgProvider {
+
+  /**
+    * Asynchronously generates a CPG from the input set of files.
+    *
+    * Depending on the implementation, this operation may fail if any of the
+    * specified files do not exist.
+    *
+    * @param fileNames a set of file names to construct the CPG from.
+    * @return A UUID identifying the eventual CPG.
+    */
+  def createCpg(fileNames: Set[String]): IO[UUID]
+
+  /**
+    * Returns a CPG result identified by `uuid` iff the CPG has been created.
+    *
+    * @param cpgId A UUID associated with a CPG.
+    * @return An [[OptionT]], containing a CPG result if it exists.
+    */
+  def retrieveCpg(cpgId: UUID): OptionT[IO, CpgOperationResult[Cpg]]
```

Comment:
Seems like a misnomer, I think, because one would except an entire CPG to be returned, while really, it just informs us about the current state the CPG is in.

Reply:
It's true that currently we only use this to determine whether or not the CPG is ready, however, it could be used by another component of the server that operates over the CPG, I'm trying to keep things in the `CpgProvider` as general as possible.